$
\renewcommand{\R}[1][{}]{{\mathbb{R}}^{#1}}
\renewcommand{\Z}[1][{}]{{\mathbb{Z}}^{#1}}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\renewcommand{\getitemR}[1]{\getItem{#1}}
\renewcommand{\getItem}{\pmb{\mid}}
\renewcommand{\elemR}[2]{{#1}^{\phantom{\T}}_{\getitemR{#2}}} 
\renewcommand{\elemRP}[2]{{\big(#1\big)}^{\phantom{\T}}_{\getitemR{#2}}} 
\renewcommand{\elemRPE}[2]{\big({#1}^{\phantom{\T}}_{\getitemR{#2}}\big)}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\renewcommand{\Vect}[2][{}]{{\boldsymbol{#2}}_{#1}}
\renewcommand{\eleVR}[2]  {\elemR  {\Vect{#1}}{#2}}	 % con subindices
\renewcommand{\eleVRP}[2] {\elemRP {\Vect{#1}}{#2}}      % con subindices y paréntesis interior
\renewcommand{\eleVRPE}[2]{\elemRPE{\Vect{#1}}{#2}}      % con subindices y paréntesis exterior
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\renewcommand{\VectC}[2][{}]  {\elemR  {\Mat{#2}}{#1}}   % con subindices
\renewcommand{\VectCP}[2][{}] {\elemRP {\Mat{#2}}{#1}}   % con subindices y paréntesis
\renewcommand{\VectCPE}[2][{}]{\elemRPE{\Mat{#2}}{#1}}   % con subindices y paréntesis exterior
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\renewcommand{\mat}[1]{\mathbf{#1}}
\renewcommand{\Mat} [2][{}]{{\mat{#2}}_{#1}} 
\renewcommand{\T}{\intercal}
\renewcommand{\MatT}[2][{}]{{\mat{#2}}^{\T}_{#1}}
\renewcommand{\VectC}[2][{}]    {\elemR  {\Mat{#2}}{#1}}   % con subindices
\renewcommand{\VectCP}[2][{}]   {\elemRP  {\Mat{#2}}{#1}}   % con subindices y paréntesis
\renewcommand{\VectCPE}[2][{}]  {\elemRPE  {\Mat{#2}}{#1}}   % con subindices y paréntesis exterior
\renewcommand{\VectCC}[2][{}]   {\elemRR {\Mat{#2}}{#1}}   % con ()
\renewcommand{\VectCCC}[2][{}]  {\elemRRR{\Mat{#2}}{#1}}   % con texto "col"
%SELECCIÓNA de FILAS y COlUMNAS DE UNA MATRIZ TRANSPUESTA PARA GENERAR UN VECTOR DE Rn
\renewcommand{\VectTC}[2][{}]   {\elemR{\MatT{#2}\!}{#1}}  % con subindices
\renewcommand{\VectTCC}[2][{}]  {\elemRR{ \MatT{#2}}{#1}}  % con ()
\renewcommand{\VectTCCC}[2][{}] {\elemRRR{\MatT{#2}}{#1}}  % con texto "col"
$

${\huge\text{Jupyter notebook de la Lección 4}}$

**Índice de contenidos de este notebook** (Requiere ejecutar todo el Notebook para que aparezca el índice)
<div id="toc"></div>

In [1]:
from notacion import *

# Transformaciones elementales

Consideraremos dos tipos de transformaciones elementales que
denotaremos con tuplas. Para indicar que una tupla es una abreviatura
de una transformación elemental, la escribiremos como argumento de la
clase `T`.

__Tipo I:__ Las transformaciones elementales del primer tipo suman $a$
veces el vector $i$-ésimo al vector $j$-ésimo:

$$
\underset{\left[\left(a\right)\mathbf{i}+\mathbf{j}
\right]}{\mathbf{\tau}}
$$

denotaremos estas transformaciones con una tupla de tres componentes:
`T( (a, i, j) )`.

Por ejemplo, la abreviatura de la transformación elemental que suma 3
veces el vector segundo al vector primero se escribe

In [2]:
T( (3, 2, 1) )

T((3, 2, 1))

Fíjese que el índice del vector modificado aparece en último lugar;
que a dicho vector se le suma un múltiplo de otro vector, cuyo índice
apare en segundo lugar; y que el coeficiente por el que se multiplica
aparece en primer lugar.

__Tipo II:__ Las Transformaciones Elementales de segundo tipo
multiplican el vector $j$-ésimo por un número $a$.

$$
\underset{\left[\left(a\right)\mathbf{j}\right]}{\mathbf{\tau}}
$$

denotaremos estas transformaciones con una tupla de dos componentes:
`T( (a, j) )`.

Por ejemplo, la transformación elemental que multiplica por 5 al
segundo vector se escribe

In [3]:
T( (5, 2) )

T((5, 2))

Nuevamente, el índice del vector modificado aparece en último lugar, y
el coeficiente por el que se multiplica aparece en primer lugar.

## Sucesión de transformaciones elementales

Es posible combinar una sucesión de transformaciones elementales; por
ejemplo una combinación de las dos transformaciones anteriores
$$
\underset{\begin{subarray}{c} \left[\left(a\right)\mathbf
{j}\right]\\\left[\left(b\right)\mathbf{p}+\mathbf{q} \right]\end
{subarray}}{\mathbf{\tau}}
$$
en este caso, el argumento de la clase `T` debe ser una lista de
abreviaturas.

Por ejemplo, podemos combinar las dos transformaciones de más arriba

In [4]:
T( [(3, 2, 1), (5, 2)] )

T([(3, 2, 1), (5, 2)])

# Transformaciones elementales de las filas o las columnas de una matriz
## Transformaciones de las columnas

Cuando aplicamos una transformación elemental por la derecha de una
matriz, modificamos sus columnas. La actuación de una transformación
elemental se indica con el símbolo `&`

Por ejemplo, considere la matriz $\Mat{A}$ 

In [5]:
A=Matrix([Vector((1,2)), Vector((0,3)), Vector((-2,2))])
A

Matrix([Vector([1, 2]), Vector([0, 3]), Vector([-2, 2])])

vamos a sumar el doble de la primera columna de $\Mat{A}$ a su tercera
columna.

In [6]:
A & T( (2,1,3) )

Matrix([Vector([1, 2]), Vector([0, 3]), Vector([0, 6])])

Cuando aplicamos una transformación elemental sobre una matriz,
modificamos la matriz. Así que ahora la matriz $\Mat{A}$ ha cambiado:

In [7]:
A

Matrix([Vector([1, 2]), Vector([0, 3]), Vector([0, 6])])

Si queremos aplicar transformaciones sobre una matriz y no perder la
matriz original, necesitamos hacer una copia. Por ejemplo, definamos
la siguiente matriz $\Mat{A}$

In [8]:
x = Vector((1, 1, -2)); y = Vector((1, -2, 1)); z = Vector((-2, 1, 1))
A = Matrix([x,y,z])
A

Matrix([Vector([1, 1, -2]), Vector([1, -2, 1]), Vector([-2, 1, 1])])

y hagamos una copia (con `Matrix(A)`) a la que aplicaremos un par de
transformaciones

In [9]:
B = Matrix(A)
B & T( (2,1,3) ) & T( (5, 2) )

Matrix([Vector([1, 1, -2]), Vector([5, -10, 5]), Vector([0, 3, -3])])

## Composición de transformaciones elementales

Si aplicamos una transformación elemental a otra transformación
elemental, logramos una transformación que es la composición de
ambas. Por ejemplo

In [10]:
z = T( (2,1,3) ) & T( (5, 2) )
z

T([(2, 1, 3), (5, 2)])

si aplicamos `z` a una copia de `A`, obtendremos el mismo resultado
que obtuvimos más arriba

In [11]:
C = Matrix(A)
C & z

Matrix([Vector([1, 1, -2]), Vector([5, -10, 5]), Vector([0, 3, -3])])

Los objetos de la clase `T` poseen un atributo `t`, donde se almacena
el listado de abreviaturas. Por ejemplo, en el atributo `t` de de `z`
aparece el listado con las transformaciones aplicadas más arriba:

In [12]:
z.t

[(2, 1, 3), (5, 2)]

### Intercambios

Hay múltiples combinaciones de transformaciones elementales que logran
intercambiar de posición dos vectores. La siguiente es un ejemplo de
secuencia que intercambia los vectores primero y tercero:

In [13]:
intercambio = T( [ (-1,3), (-1,3,1), (1,1,3), (-1,3,1) ] )
Matrix(A) & intercambio

Matrix([Vector([-2, 1, 1]), Vector([1, -2, 1]), Vector([1, 1, -2])])

Afortunadamente tenemos una forma más sencilla de realizar un
intercambio. Usaremos un conjunto de dos índices como abreviatura para
denotar el intercambio de los vectores indicados en el conjunto. Por
ejemplo `T( {s, t} )` intercambia los vectores $s$-ésimo y $t$-ésimo:

$$
\underset{\left[\mathbf{s}\rightleftharpoons\mathbf{t}\right]}{\mathbf{\tau}}
$$

Así, para intercambiar las columnas primera y tercera de una copia de
`A` basta con escribir

In [14]:
Matrix(A) & T( {1,3} )

Matrix([Vector([-2, 1, 1]), Vector([1, -2, 1]), Vector([1, 1, -2])])

## Transformaciones de las filas

Para transformar las filas, las transformaciones elementales deben
actuar por la izquierda de la matriz.

Por ejemplo, podemos intercabiar las dos primeras filas `C` del
siguiente modo

In [15]:
T( {1,2} ) & C

Matrix([Vector([1, 1, -2]), Vector([-10, 5, 5]), Vector([3, 0, -3])])

Si el conjunto solo contiene un índice, el intercambio no hace nada

In [16]:
T( {1,1} ) & C

Matrix([Vector([1, 1, -2]), Vector([-10, 5, 5]), Vector([3, 0, -3])])

Tenga en cuenta que cuando actuan varias transformaciones por la
izquierda, las primeras en actuar son las más próximas a la matriz;
así, tanto

In [17]:
T({1,2}) & T((2,2)) & Matrix(A)

Matrix([Vector([2, 1, -2]), Vector([-4, 1, 1]), Vector([2, -2, 1])])

como

In [18]:
T([ {1,2}, (2,2) ]) & Matrix(A)

Matrix([Vector([2, 1, -2]), Vector([-4, 1, 1]), Vector([2, -2, 1])])

es equivalente a (y fíjese en los paréntesis)

In [19]:
T({1,2}) & ( T((2,2)) & Matrix(A) )

Matrix([Vector([2, 1, -2]), Vector([-4, 1, 1]), Vector([2, -2, 1])])

es decir, primero se multipica por 2 la segunda fila (operación dentro
del paréntesis) y luego se intercambian las dos primeras filas.

## Transformaciones de las filas y de las columnas

Si tenemos transformaciones actuando a izquierda y derecha,
modificamos las filas y las columnas a la vez

In [20]:
T( {1,2} ) & T( (2,2) ) & Matrix(A) & T( [(-3,2,3), (10,1), {2,3}] )

Matrix([Vector([20, 10, -20]), Vector([14, -5, -2]), Vector([-4, 1, 1])])

# Matrices elementales, matrices intercambio y matrices permutación

La aplicación de una _única_ transformación elemental sobre la matriz
identidad da como resultado una matriz elemental.

__Matrices elementales Tipo I__: Veamos como la aplicación de la
transformación `T(-5,1,3)` sobre las _columnas_ de la matriz identidad
de orden 4 nos da una matriz parecida a la identidad, salvo porque
aparece un -5 en la posición (1,3)

In [21]:
I(4) & T( (-5,1,3) )

Matrix([Vector([1, 0, 0, 0]), Vector([0, 1, 0, 0]), Vector([-5, 0, 1, 0]), Vector([0, 0, 0, 1])])

Pero la misma transformación elemental aplicada sobre las _filas_ de
la matriz identidad nos da otra matriz elemental distinta: la versión
transpuesta de la matriz elemental de más arriba:

In [22]:
T( (-5,1,3) ) & I(4)

Matrix([Vector([1, 0, -5, 0]), Vector([0, 1, 0, 0]), Vector([0, 0, 1, 0]), Vector([0, 0, 0, 1])])

Por tanto, las transformaciones elementales de __Tipo I__ están
asociadas a matrices elementales distintas en función de si actuan
sobre las columnas (por la derecha) o sobre las filas (por la
izquierda)

__Matrices elementales Tipo II__: Y ahora veamos como la aplicación de
la transformación `T(100,2)` sobre las _columnas_ de la matriz
identidad de orden 3 nos da una matriz parecida a la identidad, salvo
porque aparece un 100 en la segunda posición de la diagonal, i.e.,
posición (2,2)

In [23]:
I(3) & T( (100,2) )

Matrix([Vector([1, 0, 0]), Vector([0, 100, 0]), Vector([0, 0, 1])])

Las matrices elementales __Tipo II__ son simétricas, por lo que
obtenemos identica matriz elemental cuando actuamos por la izquierda
para transformar las filas de la matriz identidad

In [24]:
T( (100,2) ) & I(3)

Matrix([Vector([1, 0, 0]), Vector([0, 100, 0]), Vector([0, 0, 1])])

Tenga en cuenta que la aplicación de una _sucesión_ de
transformaciones elementales NO suele ser una matriz elemental

In [25]:
I(3) & T( (2,1,3) ) & T( (5, 2) ) & T( (-5,3,1) ) & T ( (2,1,2) )

Matrix([Vector([-9, 0, -5]), Vector([-18, 5, -10]), Vector([2, 0, 1])])

__Matrices intercambio__: La aplicación de un _único_ intercambio
entre las columnas (o entre las filas) de la matriz identidad da como
resultado una matriz intercambio.

In [26]:
I(4) & T( {2,4} )

Matrix([Vector([1, 0, 0, 0]), Vector([0, 0, 0, 1]), Vector([0, 0, 1, 0]), Vector([0, 1, 0, 0])])

In [27]:
T( {2,4} ) & I(4)

Matrix([Vector([1, 0, 0, 0]), Vector([0, 0, 0, 1]), Vector([0, 0, 1, 0]), Vector([0, 1, 0, 0])])

__Matrices permutación__: La aplicación de una _sucesión_ de
intercambios entre las columnas (y/o entre las filas) de la matriz
identidad da como resultado una reordenación de las columnas de la
matriz identidad que denominamos matriz permutación.

In [28]:
T( [ {1,2}, {3,4} ] ) & I(4) & T({3,1}) & T({2,4})

Matrix([Vector([0, 0, 0, 1]), Vector([0, 0, 1, 0]), Vector([0, 1, 0, 0]), Vector([1, 0, 0, 0])])

## Producto de una matriz por una matriz elemental

Compruebe con ejemplos que aplicar una transformación elemental sobre
las columnas de una matriz es equivalente a multiplicar dicha matriz
por la correspondiente matriz elemental; es decir, que $$
{\Mat{A}}_{\mathbf{\tau}}=\Mat{A}\big({\Mat{I}}_{\mathbf{\tau}}\big)
$$

In [29]:
# Primero aplicamos la transformación sobre las columnas
Matrix(A) & T( (2,1,3) )

Matrix([Vector([1, 1, -2]), Vector([1, -2, 1]), Vector([0, 3, -3])])

In [30]:
# y ahora multiplicamos por la derecha por la correspondiente matriz elemental
Matrix(A) * ( I(A.n) &  T((2,1,3)) )

Matrix([Vector([1, 1, -2]), Vector([1, -2, 1]), Vector([0, 3, -3])])

Compruebe que también ocurre actuado por la izquierda para transformar
las filas. Primero aplicamos la trasformación sobre las filas

In [31]:
# Primero aplicamos la transformación sobre las filas
T( (2,1,3) ) & Matrix(A) 

Matrix([Vector([1, 1, 0]), Vector([1, -2, 3]), Vector([-2, 1, -3])])

In [32]:
# y ahora multiplicamos por la izquierda por la correspondiente matriz elemental
( I(A.m) &  T((2,1,3)) ) * Matrix(A) 


Matrix([Vector([-3, 1, -2]), Vector([3, -2, 1]), Vector([0, 1, 1])])

## Las transformaciones son asociativas respecto al producto de matrices

Compruebe con ejemplos que en general
$$
{\big(\Mat{A}\Mat{B}\big)}_{\mathbf{\tau}}=\Mat{A}\big({\Mat{B}}_{\mathbf{\tau}}\big)
$$

Recuerde hacer actuar sobre una copia de las matrices originales para
que no cambien dichas matrices originales al aplicar las
transformaciones elementales

In [33]:
# primero el lado izquierdo de la igualdad
Matrix( A*B ) & T((-5,2,1))

Matrix([Vector([81, -153, 72]), Vector([-15, 30, -15]), Vector([9, -9, 0])])

In [34]:
# ahora el lado derecho de la igualdad
A * ( Matrix(B) & T((-5,2,1)))

Matrix([Vector([81, -153, 72]), Vector([-15, 30, -15]), Vector([9, -9, 0])])

Esto es cierto incluso cuando se aplica una sucesión de
transformaciones elementales (y tanto si se transforman las filas como
las columnas).

Veamos un ejemplo de trasformación de las columnas:

In [35]:
# definimos una sucesión de transformaciones elmentales
st = T([(2, 1, 3), (5, 2), (-5, 3, 1), (2, 1, 2), {2,3}])
st 

T([(2, 1, 3), (5, 2), (-5, 3, 1), (2, 1, 2), {2, 3}])

In [36]:
Matrix( A*B ) & st

Matrix([Vector([-99, 72, 27]), Vector([21, -15, -6]), Vector([-273, 294, -21])])

In [37]:
A * ( Matrix(B) & st)

Matrix([Vector([-99, 72, 27]), Vector([21, -15, -6]), Vector([-273, 294, -21])])

y ahora uno con las filas

In [38]:
st & Matrix( A*B ) 

Matrix([Vector([21, 45, 39]), Vector([-165, -225, -300]), Vector([54, 90, 99])])

In [39]:
( st & Matrix(A) ) * B

Matrix([Vector([21, 45, 39]), Vector([-165, -225, -300]), Vector([54, 90, 99])])

# Las transformaciones elementales son operadores lineales

## Propiedad asociativa

Compruebe con ejemplos que en general
$$
{\big(\lambda\Mat{B}\big)}_{\mathbf{\tau}}=\lambda\big({\Mat{B}}_{\mathbf{\tau}}\big)
$$

Recuerde hacer actuar sobre una copia de las matrices originales para
que no cambien dichas matrices originales al aplicar las
transformaciones elementales

In [40]:
# primero el lado izquierdo de la igualdad
5*( Matrix(A) & st )

Matrix([Vector([5, -70, 65]), Vector([0, 15, -15]), Vector([35, -190, 155])])

In [41]:
# y luego el lado derecho
( 5*Matrix(A) ) & st

Matrix([Vector([5, -70, 65]), Vector([0, 15, -15]), Vector([35, -190, 155])])

Compruebe que también es cierto cuando se transforman las filas

## Propiedad distributiva

Compruebe con ejemplos que en general
$$
{\big(\Mat{A}+\Mat{B}\big)}_{\mathbf{\tau}}={\Mat{A}}_{\mathbf{\tau}} + {\Mat{B}}_{\mathbf{\tau}}
$$

In [42]:
# primero el lado izquierdo de la igualdad
Matrix( A + B ) & st

Matrix([Vector([-8, -38, 46]), Vector([2, 8, -10]), Vector([14, -136, 122])])

In [43]:
# y luego el lado derecho de la igualdad
( Matrix(A) & st ) + ( Matrix(B) & st )

Matrix([Vector([-8, -38, 46]), Vector([2, 8, -10]), Vector([14, -136, 122])])

Esto también es cierto con las transformaciones de las filas

In [44]:
st & Matrix( A + B ) 

Matrix([Vector([-8, 0, -14]), Vector([66, 90, 120]), Vector([-22, -30, -40])])

In [45]:
( st & Matrix(A) ) + ( st & Matrix(B) )

Matrix([Vector([-8, 0, -14]), Vector([66, 90, 120]), Vector([-22, -30, -40])])

# Escalonando una matriz por el método de eliminación gaussiana


La función `pivote` de nuestra librería devuelve la posición del
primer componente no negativo de un vector (cuando el vector es el
único argumento).

In [46]:
pivote(Vector([0,1,0,0,1,0]))

2

Si a la función se le da como segundo argumento una posición desde la
que empezar a buscar, entonces devuelve la posición del primer
componente no negativo a partir de la posición siguiente a la
indicada.

In [47]:
pivote(Vector([0,1,0,0,1,0]), 2)

5

Cuando la función `pivote` no encuentra un componente no negativo
devuelve el valor cero

In [48]:
pivote(Vector([0,1,0,0,0,0]), 2)

0

Usando la función `pivote` y las transformaciones elementales vamos a escalonar la matriz `A`.

In [49]:
M = Matrix(A)            # Hacemos una copia de A para no perder la matriz original
r = 0                    # Contador de pivotes
for i in range(1,M.m+1): # En cada una de las filas: 
    p = pivote((i|M),r)  #   buscamos el primer componente no nulo a la derecha de la columna r 
    if p > 0:            #   si hemos econtrado alguno (si la fila no está llena de ceros por la derecha)
        r += 1           #   el contador añade un nuevo pivote a la cuenta
        M & T( {p, r} )  #   intercambiamos la columna $r$-ésima con la columna donde encontramos el pivote 
        M & T( [(Fraction(-(i|M|j),(i|M|r)), r, j) for j in range(r+1,M.n+1)] ) # con el pivote anulamos
                                                                                # todos los componentes a su derecha

M  # por último mostramos el resultado (fíjese que M ya no es una copia de A)

Matrix([Vector([1, 1, -2]), Vector([Fraction(0, 1), Fraction(-3, 1), Fraction(3, 1)]), Vector([Fraction(0, 1), Fraction(0, 1), Fraction(0, 1)])])

Los pasos dados por el algoritmo para la matriz
$
\qquad\begin{bmatrix}1&1&-2\\1&-2&1\\-2&1&1\end{bmatrix}\qquad
$
son:
$$
\begin{bmatrix}1&1&-2\\1&-2&1\\-2&1&1\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(-1\right)\mathbf{1}+\mathbf{2} \right]\\\left[\left(2\right)\mathbf{1}+\mathbf{3} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&0&0\\1&-3&3\\-2&3&-3\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(1\right)\mathbf{2}+\mathbf{3} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&0&0\\1&-3&0\\-2&3&0\end{bmatrix}
$$
(donde hemos excluido del esquema los intercambios `T({p,r})` cuando `p==r`, pues son inecesarios).

El código de más arriba está implementado en nuestra librería en la
función `GCL`, que vamos a usar con otra matriz

In [50]:
M=Matrix([[1,2,3,4,5],[0,0,-3,-5,7],[1,3,5,0,0],[0,1,0,1,0]])
GCL(M)

Matrix([Vector([1, 0, 1, 0]), Vector([Fraction(0, 1), Fraction(-3, 1), Fraction(2, 1), Fraction(0, 1)]), Vector([Fraction(0, 1), Fraction(0, 1), Fraction(1, 1), Fraction(1, 1)]), Vector([Fraction(0, 1), Fraction(0, 1), Fraction(0, 1), Fraction(25, 3)]), Vector([Fraction(0, 1), Fraction(0, 1), Fraction(0, 1), Fraction(0, 1)])])

Los pasos dados por el algoritmo para la matriz
$
\qquad\begin{bmatrix}1&2&3&4&5\\0&0&-3&-5&7\\1&3&5&0&0\\0&1&0&1&0\end{bmatrix}\qquad
$
son
$$
\begin{bmatrix}1&2&3&4&5\\0&0&-3&-5&7\\1&3&5&0&0\\0&1&0&1&0\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(-2\right)\mathbf{1}+\mathbf{2} \right]\\\left[\left(-3\right)\mathbf{1}+\mathbf{3} \right]\\\left[\left(-4\right)\mathbf{1}+\mathbf{4} \right]\\\left[\left(-5\right)\mathbf{1}+\mathbf{5} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&0&0&0&0\\0&0&-3&-5&7\\1&1&2&-4&-5\\0&1&0&1&0\end{bmatrix}\xrightarrow{\underset{\left[\mathbf{2}\rightleftharpoons\mathbf{3}\right]}{\mathbf{\tau}}}\begin{bmatrix}1&0&0&0&0\\0&-3&0&-5&7\\1&2&1&-4&-5\\0&0&1&1&0\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(\frac{-5}{3}\right)\mathbf{2}+\mathbf{4} \right]\\\left[\left(\frac{7}{3}\right)\mathbf{2}+\mathbf{5} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&0&0&0&0\\0&-3&0&0&0\\1&2&1&\frac{-22}{3}&\frac{-1}{3}\\0&0&1&1&0\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(\frac{22}{3}\right)\mathbf{3}+\mathbf{4} \right]\\\left[\left(\frac{1}{3}\right)\mathbf{3}+\mathbf{5} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&0&0&0&0\\0&-3&0&0&0\\1&2&1&0&0\\0&0&1&\frac{25}{3}&\frac{1}{3}\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(\frac{-1}{25}\right)\mathbf{4}+\mathbf{5} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&0&0&0&0\\0&-3&0&0&0\\1&2&1&0&0\\0&0&1&\frac{25}{3}&0\end{bmatrix}
$$
(donde hemos excluido del esquema no solo los intercambios
inecesarios, sino también los inecesarios pasos que eliminan ceros a
la derecha del pivote... fíjese que el algoritmo recorre la fila desde
el pivote _hasta el final_, operando incluso cuando el componente
`i,j` es un cero).

## Otras formas de aplicar la eliminación gaussina

Como se puede comprobar, el proceso de eliminación de más arriba
emplea operaciones elementales sobre las columnas "barriendo" las
filas de arriba a abajo y las columnas de izquierda a derecha.

Pero podemos proceder de otros modos, por ejemplo de abajo a arriba y
de dercha a izquierda. De manera que logramos una matriz triangular
superior, y donde las columnas de ceros acaban a la izquierda:

In [61]:
M = Matrix([[1,2,3,4,5],[0,0,-3,-5,7],[1,3,5,0,0],[0,1,0,1,0]])

In [60]:
r = 0                              # Contador de pivotes
for i in reversed(range(1,M.m+1)): # En cada una de las filas (comenzando por la ultima): 
    p = pivote(reversed(i|M), r)   #  buscamos el primer componente no nulo a la izquierda de columna r 
    if p > 0:                      #  si la fila no está llena de ceros por la izquierda
        r += 1                     #  el contador añade un nuevo pivote a la cuenta
        M & T( {M.n-p+1, M.n-r+1} )#  intercambiamos las columnas
        M & T( [ (Fraction(-(i|M|j),(i|M|(M.n-r+1))), M.n-r+1, j ) for j in reversed(range(1,M.n-r+1)) ] )
                                   #  con el pivote anulamos todos los componentes a su izquierda
                                                                            
M  # por último mostramos el resultado (fíjese que M ya no es una copia de A)

Matrix([Vector([Fraction(0, 1), Fraction(0, 1), Fraction(0, 1), Fraction(0, 1)]), Vector([Fraction(-303, 35), Fraction(0, 1), Fraction(0, 1), Fraction(0, 1)]), Vector([Fraction(5, 1), Fraction(7, 1), Fraction(0, 1), Fraction(0, 1)]), Vector([Fraction(3, 1), Fraction(-3, 1), Fraction(5, 1), Fraction(0, 1)]), Vector([4, -5, 0, 1])])

Los pasos dados en este caso han sido
$$
\begin{bmatrix}1&2&3&4&5\\0&0&-3&-5&7\\1&3&5&0&0\\0&1&0&1&0\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\mathbf{4}\rightleftharpoons\mathbf{5}\right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&2&3&5&4\\0&0&-3&7&-5\\1&3&5&0&0\\0&1&0&0&1\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(-1\right)\mathbf{5}+\mathbf{2} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&-2&3&5&4\\0&5&-3&7&-5\\1&3&5&0&0\\0&0&0&0&1\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\mathbf{3}\rightleftharpoons\mathbf{4}\right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}1&-2&5&3&4\\0&5&7&-3&-5\\1&3&0&5&0\\0&0&0&0&1\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(\frac{-3}{5}\right)\mathbf{4}+\mathbf{2} \right]\\\left[\left(\frac{-1}{5}\right)\mathbf{4}+\mathbf{1} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}\frac{2}{5}&\frac{-19}{5}&5&3&4\\\frac{3}{5}&\frac{34}{5}&7&-3&-5\\0&0&0&5&0\\0&0&0&0&1\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(\frac{-34}{35}\right)\mathbf{3}+\mathbf{2} \right]\\\left[\left(\frac{-3}{35}\right)\mathbf{3}+\mathbf{1} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}\frac{-1}{35}&\frac{-303}{35}&5&3&4\\0&0&7&-3&-5\\0&0&0&5&0\\0&0&0&0&1\end{bmatrix}\xrightarrow{\underset{\begin{subarray}{c} \left[\left(\frac{-1}{303}\right)\mathbf{2}+\mathbf{1} \right]\end{subarray}}{\mathbf{\tau}}}\begin{bmatrix}0&\frac{-303}{35}&5&3&4\\0&0&7&-3&-5\\0&0&0&5&0\\0&0&0&0&1\end{bmatrix}
$$

_Puede intentar otras formas de escalonar una matriz, por ejemplo,
evitando usar fracciones, o escalonando de derecha a izquierda, o con
transformaciones de las filas de arriba a abajo (o de abajo a arriba),
etc._


---- FIN DE LA LECCIÓN -----



Lo que sigue es sólo para crear un índice de contenidos de este notebook (y se muestra al principio de la página)

In [54]:
%%javascript
$.getScript('https://kmahelona.github.io/ipython_notebook_goodies/ipython_notebook_toc.js')
$.getScript('https://kmahelona.github.io/ipython_notebook_goodies/ipython_notebook_toc.js')

<IPython.core.display.Javascript object>